### ISSUE TITLE
Issues with video seeking

### ISSUE THREAD

**Issue by m-hilgendorf** [filmstro/filmstro_ffmpeg#8]
I realize this isn't the most maintained repo on github, but it was perfect for hacking together a project. At the moment I'm scratching my head around how VideoReader::setNextReadPosition is supposed to work, I can't get a slider to work as a seek control reliably. If you check out the VideoPlayer example the same issue is present, jumping back and forth in the video with cause the video stream to hang, in debug configuration the message "No Frame Available" is continuously printed to the terminal, meanwhile the audio stream is unaffected. 

Not sure where to start looking to fix this.

**Comment by ffAudio** [comment]
Yes, unfortunately the project I started the wrapper for was cancelled, that's why there was no work done in the last year. I would love to continue, but I also faced problems. It worked well for many videos, but for some this approach seems problematic.

The FFmpegVideoReader::DecoderThread reads the packages ahead. By calling setNextReadPosition the stream seeks to the correct audio package with the PTS matching the readPosition. Now the thread needs to read enough packages to find a video frame with the same PTS. I assume this is the culprit, that some video formats allow the packets to be too far apart from each other, or even worse that the video frame appeared before the matching audio frame.

For a start there are some settings in the FFmpegVideoReader constructor you can try to increase. But e.g. for old divX files it never worked for me.

Some thing I wanted to do was to bring it to the latest FFmpeg api version (I used the now deprecated functions). But I don't know, if that alone would solve the issue...

If you find something, let me know, I am happy to put it in or to comment if you have further questions. I just have a bit limited time right now to work myself on it, too many unfinished stuff on the table...

**Comment by m-hilgendorf** [comment]
I don't think it has to do with the packets being too far apart, one thing that jumped out to me in `void FFmpegVideoReader::DecoderThread::setCurrentPT` was this line: 

<code> 
av_seek_frame (formatContext, audioStreamIdx, readPos, 0);
<code>

where the flags are set to zero. This is fine for the audio stream but a little change made it work better for video: 

    // check to see if we're seeking backwards
    int flags = (getCurrentPTS() > pts) ? AVSEEK_FLAG_BACKWARD : 0; 

    // allow to seek to non-keyframes, to keep up with audio stream
    flags |= AVSEEK_FLAG_ANY; 

    // seek to the new position using the flags
    av_seek_frame (formatContext, audioStreamIdx, readPos, flags);

It looks pretty bad, but appears to keep the video signal from hanging similar to how the same looks in VLC Player.
